src/KubernetesClient/Kubernetes.Generic.cs

```diff
@@ -0,0 +1,1131 @@
+            string pretty = default,
+            Dictionary<string, List<string>> customHeaders = null,
+            CancellationToken cancellationToken = default) where T : IKubernetesObject
+        {
+            var result = await ReadWithHttpMessagesAsync(typeof(T), name, namespaceParameter, exact, export, pretty, customHeaders, cancellationToken);
+            return (HttpOperationResponse<T>) result;
+        }
+
+        /// <inheritdoc cref="IKubernetes"/>
+        public async Task<HttpOperationResponse> ReadWithHttpMessagesAsync(
+            Type type,
+            string name,
+            string namespaceParameter = null,
+            bool? exact = default,
+            bool? export = default,
+            string pretty = default,
+            Dictionary<string, List<string>> customHeaders = null,
+            CancellationToken cancellationToken = default)
+        {
+
+            // Tracing
+            var shouldTrace = ServiceClientTracing.IsEnabled;
+            string invocationId = null;
+            if (shouldTrace)
+            {
+                invocationId = ServiceClientTracing.NextInvocationId.ToString();
+                var tracingParameters = new Dictionary<string, object>();
+                tracingParameters.Add("exact", exact);
+                tracingParameters.Add("export", export);
+                tracingParameters.Add("name", name);
+                tracingParameters.Add("namespaceParameter", namespaceParameter);
+                tracingParameters.Add("pretty", pretty);
+                tracingParameters.Add("cancellationToken", cancellationToken);
+                ServiceClientTracing.Enter(invocationId, this, "ReadNamespacedConfigMap", tracingParameters);
+            }
+            // Construct URL
+            var url = GetUrlPathForResource(type, namespaceParameter, name);
+            var queryParameters = new List<string>();
+            if (exact != null)
+            {
+                queryParameters.Add($"exact={System.Uri.EscapeDataString(SafeJsonConvert.SerializeObject(exact, SerializationSettings).Trim('"'))}");
+            }
+            if (export != null)
+            {
+                queryParameters.Add($"export={System.Uri.EscapeDataString(SafeJsonConvert.SerializeObject(export, SerializationSettings).Trim('"'))}");
+            }
+            if (pretty != null)
+            {
+                queryParameters.Add($"pretty={System.Uri.EscapeDataString(pretty)}");
+            }
+            if (queryParameters.Count > 0)
+            {
+                url += "?" + string.Join("&", queryParameters);
+            }
+            // Create HTTP transport objects
+            var httpRequest = new HttpRequestMessage();
+            HttpResponseMessage httpResponse = null;
+            httpRequest.Method = new HttpMethod("GET");
+            httpRequest.RequestUri = new System.Uri(url);
+            // Set Headers
+
+
+            if (customHeaders != null)
```

Comment:
A bunch of this logic is repeated in multiple places, please refactor into common methods.